File: SQL Injection/VulnerableBackend.py
import mysql.connector

def login(username, password):
    try:
        # Hardcoded credentials
        conn = mysql.connector.connect(
            host="localhost",
            user="root",
            password="password",  # Vulnerable hardcoded credentials
            database="testdb"
        )
        cursor = conn.cursor()

        # Vulnerable SQL Query
        query = f"SELECT * FROM users WHERE username = '{username}' AND password = '{password}'"
        cursor.execute(query)

        results = cursor.fetchall()
        if results:
            print("Login successful!")
        else:
            print("Invalid username or password.")
    except mysql.connector.Error as err:
        print(f"Connection Error: {err}")  # Improper error handling
    finally:
        cursor.close()
        conn.close()

def register_user(username, password, email):
    try:
        conn = mysql.connector.connect(
            host="localhost",
            user="root",
            password="password",  # Vulnerable hardcoded credentials
            database="testdb"
        )
        cursor = conn.cursor()
        # Vulnerable SQL Query (SQL Injection)
        query = f"INSERT INTO users (username, password, email) VALUES ('{username}', '{password}', '{email}')"
        cursor.execute(query)
        conn.commit()
        print("User registered successfully!")
    except mysql.connector.Error as err:
        print(f"Query Error: {err}")  
    finally:
        cursor.close()
        conn.close()

def update_password(username, new_password):
    try:
        conn = mysql.connector.connect(
            host="localhost",
            user="root",
            password="password", 
            database="testdb"
        )
        cursor = conn.cursor()


        query = f"UPDATE users SET password = '{new_password}' WHERE username = '{username}'"
        cursor.execute(query)
        conn.commit()
        print("Password updated successfully!")
    except mysql.connector.Error as err:
        print(f"Query Error: {err}")  
    finally:
        cursor.close()
        conn.close()

def delete_user(username):
    try:
        conn = mysql.connector.connect(
            host="localhost",
            user="root",
            password="password",  
            database="testdb"
        )
        cursor = conn.cursor()


        query = f"DELETE FROM users WHERE username = '{username}'"
        cursor.execute(query)
        conn.commit()
        print("User deleted successfully!")
    except mysql.connector.Error as err:
        print(f"Query Error: {err}")  
    finally:
        cursor.close()
        conn.close()

if __name__ == "__main__":

    username = input("Enter username to login: ")
    password = input("Enter password: ")
    login(username, password)

    username = input("Enter username to register: ")
    password = input("Enter password: ")
    email = input("Enter email: ")
    register_user(username, password, email)

    username = input("Enter username to update password: ")
    new_password = input("Enter new password: ")
    update_password(username, new_password)


    username = input("Enter username to delete: ")
    delete_user(username)

File: SQL Injection/VulnerableBackend.cpp
#include <iostream>
#include <string>
#include <mysql/mysql.h> 

void login(std::string username, std::string password) {
    MYSQL *conn;
    conn = mysql_init(NULL);

    if (!mysql_real_connect(conn, "localhost", "root", "password", "testdb", 0, NULL, 0)) {
        std::cerr << "Connection Error: " << mysql_error(conn) << std::endl;
        return;
    }

    std::string query = "SELECT * FROM users WHERE username = '" + username + "' AND password = '" + password + "';";

    if (mysql_query(conn, query.c_str())) {
        std::cerr << "Query Error: " << mysql_error(conn) << std::endl;
    } else {
        MYSQL_RES *res = mysql_store_result(conn);
        if (res && mysql_num_rows(res) > 0) {
            std::cout << "Login successful!" << std::endl;
        } else {
            std::cout << "Invalid username or password." << std::endl;
        }
        mysql_free_result(res);
    }

    mysql_close(conn);
}

void registerUser(std::string username, std::string password, std::string email) {
    MYSQL *conn;
    conn = mysql_init(NULL);

    if (!mysql_real_connect(conn, "localhost", "root", "password", "testdb", 0, NULL, 0)) {
        std::cerr << "Connection Error: " << mysql_error(conn) << std::endl;
        return;
    }

    std::string query = "INSERT INTO users (username, password, email) VALUES ('" + username + "', '" + password + "', '" + email + "');";

    if (mysql_query(conn, query.c_str())) {
        std::cerr << "Query Error: " << mysql_error(conn) << std::endl;
    } else {
        std::cout << "User registered successfully!" << std::endl;
    }

    mysql_close(conn);
}

void updatePassword(std::string username, std::string newPassword) {
    MYSQL *conn;
    conn = mysql_init(NULL);

    if (!mysql_real_connect(conn, "localhost", "root", "password", "testdb", 0, NULL, 0)) {
        std::cerr << "Connection Error: " << mysql_error(conn) << std::endl;
        return;
    }

    std::string query = "UPDATE users SET password = '" + newPassword + "' WHERE username = '" + username + "';";

    if (mysql_query(conn, query.c_str())) {
        std::cerr << "Query Error: " << mysql_error(conn) << std::endl;
    } else {
        std::cout << "Password updated successfully!" << std::endl;
    }

    mysql_close(conn);
}

void deleteUser(std::string username) {
    MYSQL *conn;
    conn = mysql_init(NULL);

    if (!mysql_real_connect(conn, "localhost", "root", "password", "testdb", 0, NULL, 0)) {
        std::cerr << "Connection Error: " << mysql_error(conn) << std::endl;
        return;
    }

    std::string query = "DELETE FROM users WHERE username = '" + username + "';";

    if (mysql_query(conn, query.c_str())) {
        std::cerr << "Query Error: " << mysql_error(conn) << std::endl;
    } else {
        std::cout << "User deleted successfully!" << std::endl;
    }

    mysql_close(conn);
}

int main() {
    std::string username, password, email, newPassword;

    std::cout << "Enter username to login: ";
    std::cin >> username;
    std::cout << "Enter password: ";
    std::cin >> password;
    login(username, password);

    std::cout << "Register a new user: " << std::endl;
    std::cout << "Enter username: ";
    std::cin >> username;
    std::cout << "Enter password: ";
    std::cin >> password;
    std::cout << "Enter email: ";
    std::cin >> email;
    registerUser(username, password, email);

    std::cout << "Update password for user: " << std::endl;
    std::cout << "Enter username: ";
    std::cin >> username;
    std::cout << "Enter new password: ";
    std::cin >> newPassword;
    updatePassword(username, newPassword);

    std::cout << "Delete user: " << std::endl;
    std::cout << "Enter username to delete: ";
    std::cin >> username;
    deleteUser(username);

    return 0;
}
